Bottom: 257f30de8ca0e0356dbd176eadb68dbe5ab27e26
Top:    ab0042b0b446176d96f3c40a9ad37305b40bc446
Author: Sean Hefty <sean.hefty@intel.com>
Date:   2012-05-21 18:46:36 -0700

rsocket preload: Use environment variable to set QP size

Allow the user to specify the size of the send/receive queues
through environment variables: RS_SQ_SIZE, RS_RQ_SIZE.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>


---

diff --git a/src/preload.c b/src/preload.c
index 87778eb..1214bb5 100644
--- a/src/preload.c
+++ b/src/preload.c
@@ -97,6 +97,9 @@ static int (*real_fcntl)(int socket, int cmd, ... /* arg */);
 static struct index_map idm;
 static pthread_mutex_t mut = PTHREAD_MUTEX_INITIALIZER;
 
+static int sq_size;
+static int rq_size;
+
 enum fd_type {
 	fd_normal,
 	fd_rsocket
@@ -196,6 +199,19 @@ static enum fd_type fd_close(int index, int *fd)
 	return type;
 }
 
+void getenv_options(void)
+{
+	char *var;
+
+	var = getenv("RS_SQ_SIZE");
+	if (var)
+		sq_size = atoi(var);
+
+	var = getenv("RS_RQ_SIZE");
+	if (var)
+		rq_size = atoi(var);
+}
+
 static void init_preload(void)
 {
 	static int init;
@@ -231,6 +247,8 @@ static void init_preload(void)
 	real_setsockopt = dlsym(RTLD_NEXT, "setsockopt");
 	real_getsockopt = dlsym(RTLD_NEXT, "getsockopt");
 	real_fcntl = dlsym(RTLD_NEXT, "fcntl");
+
+	getenv_options();
 	init = 1;
 out:
 	pthread_mutex_unlock(&mut);
@@ -285,6 +303,18 @@ err:
 	return ret;
 }
 
+/*
+ * Use defaults on failure.
+ */
+void set_rsocket_options(int rsocket)
+{
+	if (sq_size)
+		rsetsockopt(rsocket, SOL_RDMA, RDMA_SQSIZE, &sq_size, sizeof sq_size);
+
+	if (rq_size)
+		rsetsockopt(rsocket, SOL_RDMA, RDMA_RQSIZE, &rq_size, sizeof rq_size);
+}
+
 int socket(int domain, int type, int protocol)
 {
 	int index, ret;
@@ -297,6 +327,7 @@ int socket(int domain, int type, int protocol)
 	ret = rsocket(domain, type, protocol);
 	if (ret >= 0) {
 		fd_store(index, ret, fd_rsocket);
+		set_rsocket_options(ret);
 		return index;
 	} else {
 		fd_close(index, &ret);
